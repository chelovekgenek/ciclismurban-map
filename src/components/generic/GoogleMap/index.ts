import { Marker } from "react-google-maps"
import { default as GMap } from "./GoogleMap"

export const GoogleMap = Object.assign(GMap, {
  Marker,
})
