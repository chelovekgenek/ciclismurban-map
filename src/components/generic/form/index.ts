export * from "./Input"
export * from "./Textarea"
export * from "./Upload"
export * from "./GoogleMap"
export * from "./DateTimePicker"
