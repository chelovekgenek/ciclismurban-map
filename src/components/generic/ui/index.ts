export * from "./Grid"
export * from "./Empty"
export * from "./Checkbox"
export * from "./GoogleMap"
export * from "./Spinner"
export * from "./Input"
export * from "./Button"
export * from "./Icon"
