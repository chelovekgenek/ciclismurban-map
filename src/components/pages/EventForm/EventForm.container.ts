import { connect } from "react-redux"
import { withFormik } from "formik"
import { withRouter } from "react-router-dom"

import {
  EventsCreateActions,
  getEventsFetching,
  SelectedActions,
  getSelectedRoot,
  EventsUpdateActions,
} from "store/entities/locations"
import { TAppState } from "store/entities"
import { ExposeGroup, EventModel } from "models/location"
import { validateFormik, difference } from "helpers"

import { EventFormValues } from "./EventForm.helper"
import { IProps, EventForm } from "./EventForm"

export type IStateProps = {
  events: {
    fetching: ReturnType<typeof getEventsFetching>
  }
  selected: ReturnType<typeof getSelectedRoot>
}

export interface IDispatchProps {
  create: typeof EventsCreateActions.request
  update: typeof EventsUpdateActions.request
  getSelected: typeof SelectedActions.requestGet
  clearSelected: typeof SelectedActions.clear
}

const hocConnect = connect<IStateProps, IDispatchProps, IProps, TAppState>(
  state => ({
    events: {
      fetching: getEventsFetching(state),
    },
    selected: getSelectedRoot(state),
  }),
  {
    create: EventsCreateActions.request,
    update: EventsUpdateActions.request,
    getSelected: SelectedActions.requestGet,
    clearSelected: SelectedActions.clear,
  },
)
const hocWithFormik = withFormik<IProps, EventFormValues>({
  enableReinitialize: true,
  mapPropsToValues: props => new EventFormValues(props.selected.data as EventModel),
  handleSubmit: (values, { props }) => {
    const { data: selected } = props.selected
    if (selected) {
      props.update({ uuid: selected.uuid, payload: difference(values, selected) })
    } else {
      props.create(values)
    }
  },
  validate: v => {
    let errors = validateFormik(EventModel, [ExposeGroup.WRITE], ["image"])(v)
    if (!v.image) {
      errors.image = "Image must be provided"
    }
    return errors
  },
  validateOnBlur: true,
})

export const EventFormContainer = withRouter(hocConnect(hocWithFormik(EventForm)))
