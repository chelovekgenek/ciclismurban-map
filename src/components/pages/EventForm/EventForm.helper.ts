import moment from "moment"

import { MapOptions } from "helpers"
import { EventModel } from "models/location"

export class EventFormValues extends EventModel {
  constructor(props?: EventModel) {
    super()

    if (props) {
      this.title = props.title
      this.description = props.description
      this.point = props.point
      this.image = props.image
      this.startedAt = props.startedAt
    }
  }

  title = ""
  description = ""
  point = MapOptions.position
  image = ""
  startedAt = moment().toISOString()
}
