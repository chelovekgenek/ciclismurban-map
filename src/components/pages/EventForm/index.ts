export { EventFormContainer as EventForm } from "./EventForm.container"
