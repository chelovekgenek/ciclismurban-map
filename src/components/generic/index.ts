import * as Layout from "./layout"
import * as UI from "./ui"

export { Layout, UI }
