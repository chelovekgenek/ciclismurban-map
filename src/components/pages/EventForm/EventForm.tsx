import React, { useEffect, useMemo } from "react"
import { Form, FormikProps } from "formik"
import { RouteComponentProps } from "react-router-dom"
import { isEmpty } from "lodash-es"

import { App } from "components/generic/layout"
import { Input, Textarea, GoogleMap, DateTimePicker, Upload } from "components/generic/form"
import { Button } from "components/generic/ui"

import { IStateProps, IDispatchProps } from "./EventForm.container"
import { EventFormValues } from "./EventForm.helper"

interface RouteParams {
  id: string
}
export interface IProps extends IStateProps, IDispatchProps, RouteComponentProps<RouteParams> {}

export const EventForm: React.FC<IProps & FormikProps<EventFormValues>> = ({
  submitForm,
  resetForm,
  isValid,
  touched,
  match,
  selected,
  getSelected,
  clearSelected,
  events,
}) => {
  useEffect(() => {
    console.log("privet", match.params.id)
    if (match.params.id) {
      getSelected({ entity: "events", uuid: match.params.id })
    }
    return () => {
      clearSelected()
    }
  }, [match.params.id, getSelected, clearSelected])

  const actions = useMemo(() => {
    const reset = (
      <Button key="1" onClick={() => resetForm()} disabled={isEmpty(touched)}>
        Сбросить
      </Button>
    )
    const submit = (
      <Button key="2" type="primary" onClick={submitForm} disabled={!isValid} loading={events.fetching}>
        Сохранить
      </Button>
    )
    return selected.data ? [reset, submit] : [submit]
  }, [selected.data, events.fetching, isValid, submitForm, resetForm, touched])
  const title = useMemo(() => (selected.data ? "Редактировать событие" : "Новое событие"), [selected.data])

  return (
    <App
      content={{
        useHeader: true,
        useLayout: true,
        title,
        actions,
      }}
    >
      <Form>
        <Input label="Название" name="title" />
        <Textarea label="Описание" name="description" autosize={{ minRows: 1, maxRows: 4 }} />
        <DateTimePicker label="Дата" name="startedAt" />
        <Upload label="Изображение" name="image" />
        <GoogleMap name="point" />
      </Form>
    </App>
  )
}
