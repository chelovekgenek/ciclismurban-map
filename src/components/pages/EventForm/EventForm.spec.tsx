import React from "react"
import { shallow } from "enzyme"

import { Button } from "components/generic/ui"
import formProps from "mocks/formik-form-props"
import events from "mocks/events.json"

import { EventForm } from "./EventForm"
import { EventFormValues } from "./EventForm.helper"

describe("pages/EventForm", () => {
  const initialValues = new EventFormValues()
  const render = (props: Partial<React.ComponentProps<typeof EventForm>> = {}) =>
    shallow(
      <EventForm
        {...formProps}
        initialValues={initialValues}
        values={initialValues}
        events={{ fetching: false }}
        selected={{ fetching: false }}
        getSelected={jest.fn() as any}
        clearSelected={jest.fn() as any}
        create={jest.fn() as any}
        update={jest.fn() as any}
        match={{ params: {} } as any}
        history={{} as any}
        location={{} as any}
        {...props}
      />,
    )
  it("should match snapshot", () => {
    expect(render()).toMatchSnapshot()
  })
  it("should match snapshot if `selected.data` is provided", () => {
    expect(render({ selected: { data: events[0] } as any })).toMatchSnapshot()
  })
  it("should retreive dataset by `match.params.id`", () => {
    // TODO https://github.com/airbnb/enzyme/issues/2091
    // const props = {
    //   match: { params: { id: "id" } } as any,
    //   getSelected: jest.fn() as any,
    // }
    // render(props)
    // expect(props.getSelected).toHaveBeenCalledWith({ entity: "events" })
  })
  it("should be able to reset form if touched", () => {
    const wrapper = render({ selected: { data: events[0] } as any, touched: { title: true } })
    expect((wrapper.find("App").prop("content") as any).actions[0].props.disabled).toBe(false)
  })
})
