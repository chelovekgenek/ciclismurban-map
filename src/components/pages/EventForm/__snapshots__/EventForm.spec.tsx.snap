// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/EventForm should match snapshot 1`] = `
<App
  content={
    Object {
      "actions": Array [
        <Button
          block={false}
          disabled={true}
          ghost={false}
          htmlType="button"
          loading={false}
          onClick={[MockFunction]}
          type="primary"
        >
          Сохранить
        </Button>,
      ],
      "title": "Новое событие",
      "useHeader": true,
      "useLayout": true,
    }
  }
>
  <Form>
    <Input
      label="Название"
      name="title"
    />
    <Textarea
      autosize={
        Object {
          "maxRows": 4,
          "minRows": 1,
        }
      }
      label="Описание"
      name="description"
    />
    <DateTimePicker
      label="Дата"
      name="startedAt"
    />
    <Upload
      label="Изображение"
      name="image"
    />
    <GoogleMap
      name="point"
    />
  </Form>
</App>
`;

exports[`pages/EventForm should match snapshot if \`selected.data\` is provided 1`] = `
<App
  content={
    Object {
      "actions": Array [
        <Button
          block={false}
          disabled={true}
          ghost={false}
          htmlType="button"
          loading={false}
          onClick={[Function]}
        >
          Сбросить
        </Button>,
        <Button
          block={false}
          disabled={true}
          ghost={false}
          htmlType="button"
          loading={false}
          onClick={[MockFunction]}
          type="primary"
        >
          Сохранить
        </Button>,
      ],
      "title": "Редактировать событие",
      "useHeader": true,
      "useLayout": true,
    }
  }
>
  <Form>
    <Input
      label="Название"
      name="title"
    />
    <Textarea
      autosize={
        Object {
          "maxRows": 4,
          "minRows": 1,
        }
      }
      label="Описание"
      name="description"
    />
    <DateTimePicker
      label="Дата"
      name="startedAt"
    />
    <Upload
      label="Изображение"
      name="image"
    />
    <GoogleMap
      name="point"
    />
  </Form>
</App>
`;
